docs/develop/typescript/message-passing.mdx

```diff
@@ -196,107 +290,319 @@ await client.workflow.signalWithStart(yourWorkflow, {
 
-Queries are handled by your Workflow.
 
-Don't include any logic that causes [Command](/workflows#command) generation within a Query handler (such as executing Activities).
-Including such logic causes unexpected behavior.
+:::info NON-TYPE SAFE API CALLS
 
-Use [`handleQuery`](https://typescript.temporal.io/api/interfaces/workflow.WorkflowInboundCallsInterceptor/#handlequery) to handle Queries inside a Workflow.
+In real-world development, sometimes you may be unable to import message type objects defined by `defineQuery`, `defineSignal`, or `defineUpdate`.
+When you don't have access to the Workflow Definition or it isn't written in Typescript, you can still use non-type safe APIs and dynamic method invocation.
+Pass message type names instead of message type objects to:
 
-You make a Query with `handle.query(query, ...args)`. A Query needs a return value, but can also take arguments.
+- [`client.workflow.start`](https://typescript.temporal.io/api/classes/client.WorkflowClient#start)
+- [`WorkflowHandle.query`](https://typescript.temporal.io/api/interfaces/client.WorkflowHandle/#query)
+- [WorkflowHandle.signal](https://typescript.temporal.io/api/interfaces/client.WorkflowHandle#signal)
+- [`WorkflowHandle.executeUpdate`](https://typescript.temporal.io/api/interfaces/client.WorkflowHandle/#executeUpdate)
+- [`WorkflowHandle.startUpdate`](https://typescript.temporal.io/api/interfaces/client.WorkflowHandle/#startUpdate)
 
-<!--SNIPSTART typescript-handle-query -->
+Pass Workflow IDs to these APIs to get Workflow handles:
 
-[state/src/workflows.ts](https://github.com/temporalio/samples-typescript/blob/main/state/src/workflows.ts)
+- [`client.workflow.getHandle`](https://typescript.temporal.io/api/classes/client.WorkflowClient#gethandle)
+- [`getExternalWorkflowHandle`](https://typescript.temporal.io/api/namespaces/workflow#getExternalWorkflowHandle).
 
-```ts
-export async function trackState(): Promise<void> {
-  const state = new Map<string, number>();
-  setHandler(setValueSignal, (key, value) => void state.set(key, value));
-  setHandler(getValueQuery, (key) => state.get(key));
-  await CancellationScope.current().cancelRequested;
+:::
+
+
+## Message handler patterns {#message-handler-patterns}
+
+This section covers common write operations, such as Signal and Update handlers.
+It doesn't apply to pure read operations, like Queries or Update Validators.
+
+:::tip
+
+For additional information, see [Inject work into the main Workflow](/encyclopedia/workflow-message-passing#injecting-work-into-main-workflow), [Ensuring your messages are processed exactly once](/encyclopedia/workflow-message-passing#exactly-once-message-processing), and [this sample](https://github.com/temporalio/samples-typescript/blob/main/pdates-and-signals/safe-message-handlers/README.md) demonstrating safe `async` message handling.
```

Comment:
nit: I think that would read better as a bullet list…

Reply:
I think it's OK for now in rendered form

<img width="992" alt="image" src="https://github.com/user-attachments/assets/9f4e4bd0-dbf5-4e5f-81fb-d917199a2ede">